File: backend/app/schemas/trace.py
"""
Pydantic v2 schemas for Distributed Tracing Platform.
"""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class SpanResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    trace_id: str
    span_id: str
    parent_span_id: str | None = None
    service_name: str
    operation_name: str
    span_kind: str = "SERVER"
    status_code: str = "OK"
    duration_ms: float
    start_time: datetime
    end_time: datetime
    attributes_json: dict[str, Any] = Field(default_factory=dict)
    events_json: list[Any] = Field(default_factory=list)


class TraceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    trace_id: str
    name: str
    root_service: str
    http_method: str = "GET"
    http_status: int = 200
    duration_ms: float
    status: str = "ok"
    span_count: int = 1
    created_at: datetime
    spans: list[SpanResponse] = Field(default_factory=list)
    ai_analysis_json: dict[str, Any] | None = None


class TraceListResponse(BaseModel):
    items: list[TraceResponse]
    total: int
    page: int
    size: int
    pages: int


class ServiceNode(BaseModel):
    id: str
    label: str
    type: str = "service"  # service, database, cache, gateway, external
    status: str = "healthy"  # healthy, warning, critical
    avg_latency_ms: float
    rps: float
    error_rate_percent: float


class ServiceEdge(BaseModel):
    source: str
    target: str
    call_count: int
    avg_duration_ms: float
    error_rate_percent: float


class ServiceMapResponse(BaseModel):
    nodes: list[ServiceNode]
    edges: list[ServiceEdge]


class ServiceMetricsResponse(BaseModel):
    service_name: str
    avg_latency_ms: float
    p95_latency_ms: float
    p99_latency_ms: float
    requests_per_second: float
    error_rate_percent: float
    dependencies: list[str]
    ai_summary: str


class TraceAIAnalysisResponse(BaseModel):
    trace_id: str
    bottleneck_detected: bool
    slowest_service: str
    root_cause: str
    latency_breakdown: dict[str, float]
    optimization_suggestions: list[str]
    retry_recommendations: list[str]
    scaling_suggestions: list[str]
    performance_score: float  # 0 to 100
    confidence_score: float  # 0.0 to 1.0

File: frontend/src/types/trace.ts
/**
 * TypeScript Type Definitions for Distributed Tracing Platform
 */

export interface Span {
  id: string;
  trace_id: string;
  span_id: string;
  parent_span_id?: string;
  service_name: string;
  operation_name: string;
  span_kind: "SERVER" | "CLIENT" | "INTERNAL" | "PRODUCER" | "CONSUMER";
  status_code: "OK" | "ERROR";
  duration_ms: number;
  start_time: string;
  end_time: string;
  attributes_json: Record<string, any>;
  events_json: any[];
}

export interface Trace {
  id: string;
  trace_id: string;
  name: string;
  root_service: string;
  http_method: string;
  http_status: number;
  duration_ms: number;
  status: "ok" | "error";
  span_count: number;
  created_at: string;
  spans: Span[];
  ai_analysis_json?: TraceAIAnalysis;
}

export interface TraceListResponse {
  items: Trace[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface ServiceNode {
  id: string;
  label: string;
  type: "service" | "database" | "cache" | "gateway" | "external";
  status: "healthy" | "warning" | "critical";
  avg_latency_ms: number;
  rps: number;
  error_rate_percent: number;
}

export interface ServiceEdge {
  source: string;
  target: string;
  call_count: number;
  avg_duration_ms: number;
  error_rate_percent: number;
}

export interface ServiceMapResponse {
  nodes: ServiceNode[];
  edges: ServiceEdge[];
}

export interface ServiceMetrics {
  service_name: string;
  avg_latency_ms: number;
  p95_latency_ms: number;
  p99_latency_ms: number;
  requests_per_second: number;
  error_rate_percent: number;
  dependencies: string[];
  ai_summary: string;
}

export interface TraceAIAnalysis {
  trace_id: string;
  bottleneck_detected: boolean;
  slowest_service: string;
  root_cause: string;
  latency_breakdown: Record<string, number>;
  optimization_suggestions: string[];
  retry_recommendations: string[];
  scaling_suggestions: string[];
  performance_score: number;
  confidence_score: number;
}
